### PR TITLE
build(lint): throw error when number of extra args doesn't match the number of format specifiers in logger. 

### PR DESCRIPTION
**Pull request by Hweinstock**
Builds off of:
https://github.com/aws/aws-toolkit-vscode/pull/5878

## Problem
currently `logger.error(msg, err)` silently drops `err` if it doesn't map to a `%x` format specifier in msg.
we want to either prevent that mistake, or enhance our logger so that it prints useful parts of err instead of silently dropping it.

Also, there are cases like:
`getLogger().error('Certificate path {0} already exists', certPath)`, but this silently drops the arg `certPath`. Pretty sure this syntax is from Java IIRC and not supported by Node. 

## Solution
Provide a custom lint rule to ensure the number of extra args matches the number of format specifiers. 

### Note About Alternatives
- As an alternative, we could have transformed the extra `meta` items to strings and appended them to the log message or add them as separate log messages. This would allow usage of the form `getLogger().error('hit error', e)`.
-  If we add separate log messages, we break the expectation that each `logger.error` statement corresponds to one message in the log. 
- Conversely, if we append to the log message, this is somewhat of a hidden feature not exposed to the consumer. By forcing them to add a '%O' format specifier, we are making them clearly aware of how the error (or other object) will show up in the log, ensuring the effect does not "surprise" the user. 

## Notes
Relevant: code used to format errors: 
https://github.com/aws/aws-toolkit-vscode/blob/f806ae04f5d56b872282e42d5f765b2b1a9e28dd/packages/core/src/shared/errors.ts#L329-L334
which should now actually be used since errors won't be dropped. 

- JSCPD is failing due to some duplicates in the files touched. (none introduced here). 

---

License: I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/core/src/lambda/commands/deploySamApplication.ts [hunk 1]
```diff
@@ -0,0 +1,326 @@
+/*!
```
Comment (jpinkney-aws) [inline packages/core/src/lambda/commands/deploySamApplication.ts:1]: is this file related somehow?

**Review by jpinkney-aws — COMMENTED** [review]
It looks like theres a couple more linting that fail at the bottom of this PR as well

File: packages/core/src/lambda/commands/deploySamApplication.ts (hunk 1 above)
Comment (Hweinstock) [reply packages/core/src/lambda/commands/deploySamApplication.ts:1]: looks like some more cases got merged on that last master merge that I missed. Should be resolved now.

Comment (jpinkney-aws) [reply packages/core/src/lambda/commands/deploySamApplication.ts:1]: Oh weird, it looks like its still here. Maybe github UI is glitched? I think the PR is ready to be merged other than that

File: packages/core/src/ssmDocument/commands/deleteDocument.ts [hunk 2]
```diff
@@ -60,7 +60,7 @@ export async function deleteDocument(node: DocumentItemNodeWriteable) {
     } catch (err) {
         result = 'Failed'
         const error = err as Error
-        logger.error('Error on deleting document: %0', error)
+        logger.error('Error on deleting document: %O', error)
```
Comment (justinmk3) [inline packages/core/src/ssmDocument/commands/deleteDocument.ts:63]: Nice! This rule is finding real bugs.

File: plugins/eslint-plugin-aws-toolkits/index.ts [hunk 3]
```diff
@@ -19,6 +20,7 @@ const rules = {
     'no-string-exec-for-child-process': NoStringExecForChildProcess,
     'no-console-log': NoConsoleLog,
     'no-json-stringify-in-log': noJsonStringifyInLog,
+    'no-string-sub-mismatch': noStringSubMismatch,
```
Comment (justinmk3) [inline plugins/eslint-plugin-aws-toolkits/index.ts:23]: "printf" might be a slightly more recognizable name. Or something like that. "string substitution" sounds much broader.
```suggestion
    'no-printf-mismatch': noStringSubMismatch,
```

File: plugins/eslint-plugin-aws-toolkits/lib/rules/no-string-sub-mismatch.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+function countSubTokens(literalNode: TSESTree.StringLiteral) {
+    const formatRegExp = /%[scdjifoO%]/g
+    return literalNode.value.match(formatRegExp)?.length || 0
+}
+/**
+ * Form the error message using templates or actual values.
+ * Allows us to avoid copy pasting message into test file.
+ */
+export function formErrorMsg(substitutionTokens: string | number, numArgs: string | number): string {
+    return `String substitution has ${substitutionTokens} format specifiers, but ${numArgs} arguments.`
```
Comment (justinmk3) [inline plugins/eslint-plugin-aws-toolkits/lib/rules/no-string-sub-mismatch.ts:21]:
```suggestion
    return `printf-style (console.log) call has ${substitutionTokens} format specifiers, but ${numArgs} arguments.`
```

File: packages/core/src/shared/filesystemUtilities.ts [hunk 5]
```diff
@@ -245,7 +245,7 @@ export async function setDefaultDownloadPath(downloadPath: string) {
             await globals.globalState.update('aws.downloadPath', path.dirname(downloadPath))
         }
     } catch (err) {
-        getLogger().error('Error while setting "aws.downloadPath"', err as Error)
+        getLogger().error('Error while setting "aws.downloadPath" \n %O', err as Error)
```
Comment (justinmk3) [inline packages/core/src/shared/filesystemUtilities.ts:248]: nit: it's usually quite readable to start the object at the end of the line, it will look like:

```
error: ... {
  ...
}
```
```suggestion
        getLogger().error('Error while setting "aws.downloadPath": %O', err as Error)
```

File: packages/core/src/lambda/commands/deploySamApplication.ts (hunk 1 above)
Comment (justinmk3) [reply packages/core/src/lambda/commands/deploySamApplication.ts:1]: Try `git rm <filepath>` on this file then push to PR.

**Review by justinmk3 — APPROVED** [review]

File: packages/core/src/test/awsService/appBuilder/samProject.test.ts [hunk 6]
```diff
@@ -78,7 +78,7 @@ describe('samProject', () => {
             // simulate error when no samconfig.toml file in directory
             const result = await getStackName(projectRoot)
             assert.deepStrictEqual(result, {})
-            assertLogsContain('No stack name or region information available in samconfig.toml', true, 'info')
+            assertLogsContain('No stack name or region information available in samconfig.toml: %O', true, 'info')
```
[the remark this answers is not in the document]
Comment (Hweinstock) [reply packages/core/src/test/awsService/appBuilder/samProject.test.ts:81]: Yes, but believe it is only because we are in the `TestLogger`, which adds the raw strings to an array. (https://github.com/aws/aws-toolkit-vscode/blob/b2ebc10342f63ac95cd3055393e715cb369ec450/packages/core/src/test/testLogger.ts)

**Review by justinmk3 — APPROVED** [review]
